File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/lint/JmlLintingFacade.java
package com.github.jmlparser.lint;

import com.github.javaparser.ast.Node;
import com.github.jmlparser.lint.sarif.*;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.Exception;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;

/**
 * @author Alexander Weigl
 * @version 1 (12/29/21)
 */
public class JmlLintingFacade {
    private static final Logger LOGGER = LoggerFactory.getLogger(JmlLintingFacade.class);
    private static final String VERSION = JmlLintingFacade.class.getPackage().getImplementationVersion();
    private static final String NAME = "JML-lint";

    @Getter
    private final List<LintRule> linters;
    private final JmlLintingConfig config;

    public JmlLintingFacade(JmlLintingConfig config) {
        linters = getLinter(config);
        this.config = config;
    }

    private Tool getSarifTool() {
        return new Tool(
                new ToolComponent().withVersion(VERSION).withName(NAME)
                        .withShortDescription(new MultiformatMessageString().withText("Linting for the Java Modeling Language")),
                linters.stream().map(it -> new ToolComponent().withFullName(it.getClass().getName())).collect(Collectors.toSet()),
                new PropertyBag()
        );
    }


    private static List<LintRule> getLinter(JmlLintingConfig config) {
        ServiceLoader<LintRule> loader = ServiceLoader.load(LintRule.class);
        List<LintRule> validators = new ArrayList<>(64);
        for (LintRule lintRule : loader) {
            if (!config.isDisabled(lintRule)) {
                validators.add(lintRule);
            }
        }
        return validators;
    }

    public void lint(LintProblemReporter reporter, Collection<? extends Node> nodes) {
        for (Node it : nodes) {
            for (LintRule linter : linters) {
                try {
                    linter.accept(it, reporter, config);
                } catch (Exception e) {
                    LOGGER.error("Error in linter: {}", linter.getClass().getName(), e);
                }
            }
        }
    }

    public Collection<LintProblem> lint(@NotNull Collection<? extends Node> nodes) {
        var problems = new ArrayList<LintProblem>(1024);
        Consumer<LintProblem> collector = problems::add;
        lint(new LintProblemReporter(collector), nodes);
        problems.sort(Comparator.comparing(it -> it.location().toRange().get().begin));
        return problems;
    }

    public SarifSchema asSarif(Collection<LintProblem> problems) throws URISyntaxException {
        List<Result> results = problems.stream().map(this::asSarif).toList();
        List<Run> runs = List.of(new Run().withTool(getSarifTool()).withResults(results));
        return new SarifSchema(
                new URI("http://json.schemastore.org/sarif-2.1.0-rtm.4"),
                "2.1.0",
                runs, Set.of(), new PropertyBag());
    }

    private Result asSarif(LintProblem it) {
        return new Result().withRuleId(it.ruleId()).withKind(it.category()).withLevel(it.level())
                .withLocations(List.of(new Location())).withMessage(new Message().withText(it.message()));
    }
}

File: tools/lint/src/main/kotlin/io/github/jmltoolkit/lint/JmlLintingFacade.kt
package io.github.jmltoolkit.lint

import com.github.javaparser.ast.Node
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import se.bjurr.violations.lib.model.generated.sarif.*
import java.util.*
import java.util.function.Consumer
import kotlin.Exception

/**
 * @author Alexander Weigl
 * @version 1 (12/29/21)
 */
class JmlLintingFacade(private val config: JmlLintingConfig) {
    val linters: List<LintRule> = getLinter(config)

    private val sarifTool: Tool
        get() = Tool().withDriver(
            ToolComponent().withVersion(VERSION).withName(NAME)
                .withShortDescription(MultiformatMessageString().withText("Linting for the Java Modeling Language"))
                .withRules(linters.map { ReportingDescriptor().withName(it.javaClass.getName()) }.toSet())
        )


    fun lint(reporter: LintProblemReporter, nodes: Collection<Node>) {
        for (it in nodes) {
            for (linter in linters) {
                try {
                    linter.accept(it, reporter, config)
                } catch (e: Exception) {
                    LOGGER.error("Error in linter: {}", linter.javaClass.getName(), e)
                }
            }
        }
    }

    fun lint(nodes: Collection<Node>): Collection<LintProblem> {
        val problems: ArrayList<LintProblem> = ArrayList<LintProblem>(1024)
        val collector: Consumer<LintProblem> = Consumer<LintProblem> { e: LintProblem -> problems.add(e) }
        lint(LintProblemReporter(collector), nodes)
        problems.sortWith(Comparator.comparing { it.location!!.toRange().get().begin })
        return problems
    }

    fun asSarif(problems: Collection<LintProblem>): SarifSchema {
        val results = problems.map { it: LintProblem -> this.asSarif(it) }
        val runs: List<Run> = listOf(Run().withTool(sarifTool).withResults(results))
        return SarifSchema()
            .withVersion("2.1.0")
            .withRuns(runs)
    }

    private fun asSarif(it: LintProblem): Result {
        return Result().withRuleId(it.ruleId).withKind(it.category).withLevel(it.level)
            .withLocations(listOf(Location())).withMessage(Message().withText(it.message))
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(JmlLintingFacade::class.java)
        private val VERSION: String = JmlLintingFacade::class.java.getPackage().implementationVersion ?: "n/a"
        private const val NAME = "JML-lint"

        private fun getLinter(config: JmlLintingConfig): List<LintRule> {
            val loader: ServiceLoader<LintRule> = ServiceLoader.load(LintRule::class.java)
            val validators: MutableList<LintRule> = ArrayList<LintRule>(64)
            for (lintRule in loader) {
                if (!config.isDisabled(lintRule)) {
                    validators.add(lintRule)
                }
            }
            return validators
        }
    }
}
